utils/scarb-ui/src/lib.rs

```diff
@@ -149,7 +179,28 @@ impl Ui {
 
     fn do_print<T: Message>(&self, message: T) {
         match self.output_format {
-            OutputFormat::Text => message.print_text(),
+            OutputFormat::Text => {
+                let read_lock = self.state.read().expect("failed to acquire read lock");
+                if let Some(widget) = read_lock.active_print_overrider.as_ref() {
+                    match widget.println(message.text().as_str()) {
+                        PrintOverriderState::Active => {
+                            // Print has succeeded.
+                        }
+                        PrintOverriderState::Inactive => {
+                            // The driver is inactive. Print has failed.
+                            message.print_text();
+                            // Clear widget from state.
+                            drop(read_lock);
+                            self.state
+                                .write()
+                                .expect("failed to acquire write lock")
+                                .active_print_overrider = None;
+                        }
+                    }
```

Comment:
this is weird

make the `Weak` should be simply made part of public API? because you are simply creating an abstraction over abstraction that could be wrong in the first place